tests/user/test_user_added_secret_redact.py

```diff
@@ -0,0 +1,74 @@
+"""Test password sanitization in user debug logging."""
+
+import contextlib
+import io
+import re
+import unittest
+from unittest.mock import Mock, patch
+
+import __init__
+
+import tests.user.test_user_constants as TestUserConstants
+from pyracf import SecurityRequestError, UserAdmin
+from pyracf.common.irrsmo00 import IRRSMO00
+
+# Resolves F401
+__init__
+
+
+@patch("pyracf.common.irrsmo00.IRRSMO00.call_racf")
+class TestUserAddedSecretRedact(unittest.TestCase):
```

Comment:
Is there a reason why we shouldn't be doing these tests in `TestUserDebugLogging`? Also, referering to these tests as testing "redact additional secrets" instead of "added secrets redact" is more clear to me.

Reply:
The tests have to be separate and other user tests will break if this new test is not the last one run. Since it hinges on redacting a field used elsewhere in testing (`omvs:uid`), defining the object with this additional secret breaks other tests as there is no supported mechanism currently in place to remove a secret. I concede that it is entirely possible there is a better way to force this test to have a separate instance of UserAdmin that would not impact other tests, but my attempts to do so did not work.

I did change the names as indicated.

Reply:
I think I know why this is breaking your tests.

Notice how __secret_traits is initialized once here.

```python
class SecurityAdmin:
    """Base Class for RACF Administration Interface."""

    _valid_segment_traits = {}
    _common_base_traits_data_set_generic = {
        "base:aclcnt": "racf:aclcnt",
        "base:aclacnt": "racf:aclacnt",
        "base:aclacs": "racf:aclacs",
        "base:aclid": "racf:aclid",
        "base:acl2cnt": "racf:acl2cnt",
        "base:acl2acnt": "racf:acl2acnt",
        "base:acl2acs": "racf:acl2acs",
        "base:acl2cond": "racf:acl2cond",
        "base:acl2ent": "racf:acl2ent",
        "base:acl2id": "racf:acl2id",
        "base:acsaltr": "racf:acsaltr",
        "base:acscntl": "racf:acscntl",
        "base:acsread": "racf:acsread",
        "base:acsupdt": "racf:acsupdt",
        "base:all": "racf:all",
        "base:audaltr": "racf:audaltr",
        "base:audcntl": "racf:audcntl",
        "base:audnone": "racf:audnone",
        "base:audread": "racf:audread",
        "base:audupdt": "racf:audupdt",
        "base:authuser": "racf:authuser",
        "base:fvolume": "racf:fvolume",
        "base:gaudaltr": "racf:gaudaltr",
        "base:gaudcntl": "racf:gaudcntl",
        "base:gaudnone": "racf:gaudnone",
        "base:gaudread": "racf:gaudread",
        "base:gaudupdt": "racf:gaudupdt",
        "base:generic": "racf:generic",
    }

    __secret_traits = {
        "base:password": "racf:password",
        "base:passphrase": "racf:phrase",
    }

    __logger = Logger()

    def __init__(
        self,
        profile_type: str,
        debug: bool = False,
        generate_requests_only: bool = False,
        add_field_data: Union[dict, None] = None,
        overwrite_field_data: Union[dict, None] = None,
        additional_secret_traits: Union[List[str], None] = None,
    ) -> None:
        self.__irrsmo00 = IRRSMO00()
        self.__profile_type = profile_type
        self._segment_traits = {}
        # used to preserve segment traits for debug logging.
        self.__preserved_segment_traits = {}
        self._trait_map = {}
        self.__debug = debug
        self.__generate_requests_only = generate_requests_only
        if add_field_data is not None:
            self.__add_field_data(add_field_data)
        if overwrite_field_data is not None:
            self.__overwrite_field_data(overwrite_field_data)
        if additional_secret_traits is not None:
            self.__add_additional_secret_traits(additional_secret_traits)
```

I think moving the initialization of `__secret_traits` to the constructor would resolve the issue since __secret_traits would be rebuild every time you call the constructor.

It is also worth noting that the custom field data may actually lead to the same problem and that each child classes definition of `_valid_segment_traits` might be better defined as part of the constructor too.

Reply:
That makes perfect sense. I have addressed this and the unit testing works perfectly normally, so I returned the tests to `test_user_debug_logging` where they still pass the test. I have also moved the `_valid_segment_traits` definitions into the class constructors as well, though I am surprised that this didn't cause similar problems in unit testing originally.

Reply:
So, I think the general rule of thumb is that class attributes that are static should be defined directly to the class.

```python
class MyClass():
    my_static_attribute = "Hello World!"
```

And class attributes that are dynamic and or specific to a an instance of a class should be initialized in the constructor.

```python
class MyClass():
    def __init__(self):
        my_dynamic_instance_specific_attribute = "Hello World!"
```